google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionGroup.java

```diff
@@ -80,25 +83,92 @@ public void getPartitions(
+                observer.onNext(queryPartition);
+                return null;
+              }
+            });
+
+        observer.onCompleted();
       } catch (ApiException exception) {
         span.setStatus(Status.UNKNOWN.withDescription(exception.getMessage()));
         throw FirestoreException.forApiException(exception);
       } finally {
         span.end(TraceUtil.END_SPAN_OPTIONS);
       }
+    }
+  }
+
+  public ApiFuture<List<QueryPartition>> getPartitions(long desiredPartitionCount) {
+    if (desiredPartitionCount == 1) {
+      // Short circuit if the user only requested a single partition.
+      return ApiFutures.immediateFuture(
+          Collections.singletonList(new QueryPartition(partitionQuery, null, null)));
+    } else {
+      PartitionQueryRequest.Builder request = buildRequest(desiredPartitionCount);
 
-      @Nullable Object[] lastCursor = null;
-      for (Cursor cursor : response.iterateAll()) {
-        Object[] decodedCursorValue = new Object[cursor.getValuesCount()];
-        for (int i = 0; i < cursor.getValuesCount(); ++i) {
-          decodedCursorValue[i] = UserDataConverter.decodeValue(rpcContext, cursor.getValues(i));
-        }
-        observer.onNext(new QueryPartition(queryWithDefaultOrder, lastCursor, decodedCursorValue));
-        lastCursor = decodedCursorValue;
+      final TraceUtil traceUtil = TraceUtil.getInstance();
+      Span span = traceUtil.startSpan(TraceUtil.SPAN_NAME_PARTITIONQUERY);
+      try (Scope scope = traceUtil.getTracer().withSpan(span)) {
+        return ApiFutures.transform(
+            rpcContext.sendRequest(
+                request.build(), rpcContext.getClient().partitionQueryPagedCallable()),
+            new ApiFunction<PartitionQueryPagedResponse, List<QueryPartition>>() {
+              @Override
+              public List<QueryPartition> apply(PartitionQueryPagedResponse response) {
+                final ImmutableList.Builder<QueryPartition> partitions = ImmutableList.builder();
+                consumePartitions(
+                    response,
+                    new Function<QueryPartition, Void>() {
+                      @Override
+                      public Void apply(QueryPartition queryPartition) {
+                        partitions.add(queryPartition);
+                        return null;
+                      }
+                    });
+                return partitions.build();
+              }
+            },
+            MoreExecutors.directExecutor());
+      } catch (ApiException exception) {
+        span.setStatus(Status.UNKNOWN.withDescription(exception.getMessage()));
+        throw FirestoreException.forApiException(exception);
+      } finally {
+        span.end(TraceUtil.END_SPAN_OPTIONS);
       }
-      observer.onNext(new QueryPartition(queryWithDefaultOrder, lastCursor, null));
     }
+  }
 
-    observer.onCompleted();
+  private PartitionQueryRequest.Builder buildRequest(long desiredPartitionCount) {
```

Comment:
I think this could just return PartitionQueryRequest. We don't modify this builder anymore. Sorry for not noticing earlier.

Reply:
No worries, I'll clean it up before merging.